### PR TITLE
[NPUW] Fix Omni Thinker LLM with Two Outputs Issue

### PR DESCRIPTION
**Pull request by GuoliangShiIntel**
### Details:
There is a previous PR (#30535) to fix 3D position input for Omin Thinker LLM.
However, it still issue to get the right result caused by two outputs.

### Tickets:
 - [*CVS-168493*](https://jira.devtools.intel.com/browse/CVS-168493)

### CODE REVIEW

**Comment by GuoliangShiIntel** [comment]
@AsyaPronina @dmatveev Please take a look.

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 1]
```diff
@@ -347,8 +347,13 @@ std::shared_ptr<ov::Model> cvt_kvcache_to_fp16(const std::shared_ptr<ov::Model>&
...
 
 std::shared_ptr<ov::Model> redirect_new_kv_to_output(const std::shared_ptr<ov::Model>& model) {
-    const auto kStartOutputKVCacheLayers = 1u;
-    for (std::size_t i = kStartOutputKVCacheLayers; i < model->outputs().size(); ++i) {
+    const auto modelOutputSize = model->outputs().size();
+    // FIXME: This is a very tricky method. Key and value should be paired
+    // If the total output size is odd, it indicates one output. If even, it indicates two outputs
+    // Most LLMs have one output, but Qwen2.5 Omni Thinker LLM has two
+    // A more robust method is needed for general cases
+    const auto kStartOutputKVCacheLayers = (modelOutputSize % 2 == 0) ? 2u : 1u;
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:355]: Can we please try to find the index by name of the output? Just go through the outputs and stop on the needed?
Because it seems very tricky, I don't think we should merge this..

**Comment by dmatveev** [comment]
@AlexanderKalistratov can you please have a look? Thanks

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 1 above)
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: I think that first k/v layer should be determined in constructor of `LLMCompiledModel` and passed here. Just by going through output layers and finding `present_key` or something like that.

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 2]
```diff
@@ -366,5 +372,8 @@ ov::SoPtr<ov::ITensor> ov::npuw::LLMInferRequest::get_tensor(const ov::Output<co
     if (port == get_outputs()[0]) {
         return m_logits;
     }
+    if (get_outputs().size() > 1 && port == get_outputs()[1]) {
+        return m_hidden_states;
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:846]: Might be we can start to store vector of outputs in `LLMInferRequest` for matched outputs of the model.

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 1 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: @AsyaPronina Sorry for the delayed response; I was occupied with other urgent tasks.

Regarding this issue, I have updated the code to consistently find and match the output with the output name. Please review the changes. Thank you.

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 3]
```diff
@@ -592,9 +592,18 @@ std::shared_ptr<ov::Model> cvt_kvcache_to_fp16(const std::shared_ptr<ov::Model>&
...
-    for (std::size_t i = kStartOutputKVCacheLayers; i < model->outputs().size(); ++i) {
+    for (std::size_t i = 0; i < model->outputs().size(); ++i) {
         auto kvout = model->output(i);
+        auto output_name = kvout.get_any_name();
+
+        if (output_name.find("logits") != std::string::npos) {
+            continue;
+        }
+
+        if (output_name == "hidden_states") {
```
Comment (GuoliangShiIntel) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:605]: Self Review: Here, I have set "==" to indicate that the output name should be strictly matched. I am concerned that "hidden_states" might exist and potentially affect the KV cache output.

For long-term code development, I don't believe this is a sustainable approach for handling more cases in the future.

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp (hunk 2 above)
Comment (GuoliangShiIntel) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:846]: Update code with name matched

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp [hunk 4]
```diff
@@ -534,6 +534,9 @@ void ov::npuw::LLMInferRequest::infer_prefill(ov::SoPtr<ov::ITensor> input_ids,
         m_logits = m_lm_head_request->get_tensor(m_lm_head_logits_port);
     } else {
         m_logits = m_prefill_request->get_tensor(m_prefill_out_ports.at(layer_names::logits));
+        if (m_prefill_out_ports.find("hidden_states") != m_prefill_out_ports.end()) {
+            m_hidden_states = m_prefill_request->get_tensor(m_prefill_out_ports.at(layer_names::hidden_states));
+        }
     }
```
Comment (GuoliangShiIntel) [inline src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:733]: @AsyaPronina I have updated this PR to resolve the conflict. I noticed that there are two branches to obtain the `m_logits` output without the `m_lm_head_request` parameter. I'm unsure about the background of this configuration, so I did not add code to retrieve the `m_hidden_states` output when it is true. Please provide your comments.

**Comment by GuoliangShiIntel** [comment]
@AsyaPronina @AlexanderKalistratov not sure if you have other comments about this PR

File: src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp (hunk 4 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:733]: @AsyaPronina your three-model pipeline guidance is required here.

Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:733]: This `m_lm_head_request` is for the three-model pipeline, when LM head - matrix multiplication of output embeddings with vocabulary - is separated to additional 3rd model. As this 3rd model should only be connected to 1 `Result` (logits), I expect that `hidden_states` will stay in prefill/kvcache models and won't get into 3rd model. So, in the if above you should try to get `hidden_states` from the prefill as in the if below.

However, due to 2 outputs, pattern to separate 3rd model can fail matching or match another output. Better to test first, how should this work? Let us validate this setting before merge of this PR together.
